### PR TITLE
Update comparison table

### PR DESCRIPTION
**Pull request by nhooyr**
I tried my best to keep the comparison objective but please let
me know if you'd prefer anything changed.

### Gorilla WebSocket compared with other packages

This comparison used to include [golang.org/x/net/websocket](http://godoc.org/golang.org/x/net/websocket) but it has
been deprecated. Do not use [golang.org/x/net/websocket](http://godoc.org/golang.org/x/net/websocket) in future projects.

See [golang/go#18152](https://github.com/golang/go/issues/18152)

Another comparison of available Go WebSocket libraries is available at [nhooyr.io/websocket](https://github.com/nhooyr/websocket#comparison)

#### [nhooyr.io/websocket](https://nhooyr.io/websocket)

gorilla/websocket is 6 years old and thus very mature and battle tested. On the other hand, it has
accumulated cruft over the years.
[nhooyr.io/websocket](http://godoc.org/nhooyr.io/websocket) is much newer and attempts to
provide a more idiomatic, minimal and useful API for WebSockets.

|  | [github.com/gorilla](http://godoc.org/github.com/gorilla/websocket) | [nhooyr.io/websocket](http://godoc.org/nhooyr.io/websocket) |
| :--- | ---: | ---: |
| **[RFC 6455](http://tools.ietf.org/html/rfc6455) Features** | 
| Passes [Autobahn Test Suite](https://github.com/crossbario/autobahn-testsuite) | [Yes](https://github.com/gorilla/websocket/tree/master/examples/autobahn) | [Yes](https://github.com/nhooyr/websocket/blob/master/conn_test.go) |
| Receive [fragmented](https://tools.ietf.org/html/rfc6455#section-5.4) message | Yes | Yes |
| Send [close](https://tools.ietf.org/html/rfc6455#section-5.5.1) message | [Yes](http://godoc.org/github.com/gorilla/websocket#hdr-Control_Messages) | [Yes](https://godoc.org/nhooyr.io/websocket#Conn.Close) |
| Send [pings](https://tools.ietf.org/html/rfc6455#section-5.5.2) and receive [pongs](https://tools.ietf.org/html/rfc6455#section-5.5.3) | [Yes](http://godoc.org/github.com/gorilla/websocket#hdr-Control_Messages) | [Yes](https://godoc.org/nhooyr.io/websocket#Conn.Ping)
| Get [type](https://tools.ietf.org/html/rfc6455#section-5.6) of received message | Yes | Yes |
| **Other Features** |
| Well tested | Yes | [Yes](https://codecov.io/gh/nhooyr/websocket) | 
| Low level protocol control | [Yes](http://godoc.org/github.com/gorilla/websocket#hdr-Control_Messages) | No |
| [Compression Extensions](https://tools.ietf.org/html/rfc7692) | [Experimental](https://godoc.org/github.com/gorilla/websocket#hdr-Compression_EXPERIMENTAL) | [No](https://github.com/nhooyr/websocket#design-justifications) |
| Read with [io.Reader](https://golang.org/pkg/io/#Reader) | [Yes](http://godoc.org/github.com/gorilla/websocket#Conn.NextReader) | [Yes](https://godoc.org/nhooyr.io/websocket#Conn.Reader) |
| Write with [io.WriteCloser](https://golang.org/pkg/io/#WriteCloser) | [Yes](http://godoc.org/github.com/gorilla/websocket#Conn.NextWriter) | [Yes](https://godoc.org/nhooyr.io/websocket#Conn.Writer) |
| Compile for [Wasm](https://github.com/golang/go/wiki/WebAssembly) | [No](https://github.com/gorilla/websocket/issues/432) | [Yes](https://godoc.org/nhooyr.io/websocket#hdr-Wasm) |
| Use stdlib [*http.Client](https://golang.org/pkg/net/http/#Client) | No | [Yes](https://godoc.org/nhooyr.io/websocket#DialOptions) |
| Concurrent writers | [No](https://godoc.org/github.com/gorilla/websocket#hdr-Concurrency) | [Yes](https://godoc.org/nhooyr.io/websocket#Conn) |
| Configurable buffer sizes | [Yes](https://godoc.org/github.com/gorilla/websocket#hdr-Buffers) | No |
| Prepared messages | [Yes](https://godoc.org/github.com/gorilla/websocket#PreparedMessage) | No |
| Full [context.Context](https://blog.golang.org/context)  support | No | Yes
| [net.Conn](https://golang.org/pkg/net/#Conn) adapter | No | [Yes](https://godoc.org/nhooyr.io/websocket#NetConn) |

### CODE REVIEW

**Comment by nhooyr** [comment]
I've switched the comparison to also use markdown syntax to make it easier to read.

**Comment by nhooyr** [comment]
I've also added a link to this comparison from my library's README.md.

**Comment by elithrar** [comment]
Thanks for this, and understand the effort, but I don’t think this makes sense.

The /x/net/websocket comparison does as many users look to the std lib for packages, and it’s large lack of API coverage is a real gap.

**Comment by nhooyr** [comment]
> Thanks for this, and understand the effort, but I don’t think this makes sense.
> The /x/net/websocket comparison does as many users look to the std lib for packages, and it’s large lack of API coverage is a real gap.

I'd have to disagree given the x stdlib package is already effectively deprecated and points users towards gorilla/websocket. This comparison makes it seem like only gorilla/websocket and the x/net libraries are the Go WebSocket libraries available but there is mine and gobwas/ws which include additional features that would be valuable to many users.

For example, my library can compile to WASM which is a requested feature here #368 

It implements a net.Conn wrapper which many users have told me they found very useful as it makes it easy to interoperate with other networking libraries and eases the transition from golang.org/x/net/websocket.

I also think the context.Context support and concurrent writing is a big one that users of this library have to handle themselves. E.g. see #87.

Likewise, as soon as WebSockets over HTTP/2 can work with the Go stdlib, I'll be implementing them very quickly as my library uses the stdlib *http.Client directly. gorilla/websocket will have a harder time as it exposes the direct net.Conn. Likewise with the WASM support. It also means I do not have to implement things like #516.

There's also the wspb/wsjson subpackages which reuse buffers automatically.

My library also makes it simpler to deal with write only WebSocket connections and pings. See #492 for gorilla issue regarding this.

To summarize, my point is that there is real demand for the features my library implements and it's evident in the gorilla/websocket issue tracker. Rather than duplicate every feature here, it makes more sense to point users to my library if they're looking for these features. Likewise with gobwas/ws but they'll see that in my libraries's README.md so I didn't include it here but I can if you'd like.

I'm not saying everyone should switch, I've made it clear in this comparison and the one on my repo that gorilla/websocket is the mature and battle tested option and if that's what you're looking for, gorilla/webosocket is the choice.

**Comment by nhooyr** [comment]
I can clarify the comparison further if that would help.

**Comment by elithrar** [comment]
No - I understand where you are coming from.

I’m happy to have a list of alternative libraries as one-liners (and
links); I’m not interested in maintaining documentation in this package for
other libraries.

We can agree to disagree here, but this won’t be merged as-is.

**Comment by nhooyr** [comment]
I'm happy to maintain it. See my comment in #370

For now I'll follow up with a PR with just the links.

Thanks!

**Comment by elithrar** [comment]
Thanks.
